Lib/asyncio/events.py

```diff
@@ -47,60 +49,89 @@ def __init__(self, callback, args, loop, context=None):
         self._loop = loop
-        self._callback = callback
-        self._args = args
-        self._cancelled = False
+        self._callback_args = (callback, args)
         self._repr = None
         if self._loop.get_debug():
             self._source_traceback = format_helpers.extract_stack(
                 sys._getframe(1))
         else:
             self._source_traceback = None
 
-    def _repr_info(self):
+    def _repr_info(self, cancelling, callback_args):
         info = [self.__class__.__name__]
-        if self._cancelled:
+        if cancelling:
+            info.append('cancelled')
+        if callback_args is _HANDLE_CANCELLED:
             info.append('cancelled')
-        if self._callback is not None:
+            callback = None
+            args = None
+        else:
+            callback, args = callback_args
+
+        if callback is not None:
             info.append(format_helpers._format_callback_source(
-                self._callback, self._args,
+                callback, args,
                 debug=self._loop.get_debug()))
         if self._source_traceback:
             frame = self._source_traceback[-1]
             info.append(f'created at {frame[0]}:{frame[1]}')
         return info
 
+    def _repr_atomic(self, cancelling, callback_args):
+        info = self._repr_info(cancelling, callback_args)
+        return '<{}>'.format(' '.join(info))
+
     def __repr__(self):
         if self._repr is not None:
             return self._repr
-        info = self._repr_info()
-        return '<{}>'.format(' '.join(info))
+        return self._repr_atomic(cancelling=False, callback_args=self._callback_args)
 
     def get_context(self):
         return self._context
 
     def cancel(self):
-        if not self._cancelled:
-            self._cancelled = True
+        callback_args = self._callback_args
+        self._callback_args = _HANDLE_CANCELLED
+        if callback_args is not _HANDLE_CANCELLED:
             if self._loop.get_debug():
                 # Keep a representation in debug mode to keep callback and
                 # parameters. For example, to log the warning
                 # "Executing <Handle...> took 2.5 second"
-                self._repr = repr(self)
-            self._callback = None
-            self._args = None
+                self._repr = self._repr_atomic(cancelling=True, callback_args=callback_args)
 
     def cancelled(self):
-        return self._cancelled
+        return self._callback_args is _HANDLE_CANCELLED
+
+    @property
+    def _cancelled(self):
+        return self.cancelled()
+
+    @property
+    def _callback(self):
+        callback_args = self._callback_args
+        if callback_args is _HANDLE_CANCELLED:
+            return None
+        return callback_args[0]
+
+    @property
+    def _args(self):
+        callback_args = self._callback_args
+        if callback_args is _HANDLE_CANCELLED:
+            return None
+        return callback_args[1]
```

Comment:
the tests access private attributes - I'd rather update those tests. These properties were just to get a PR up and passing CI quickly.